### ISSUE TITLE
Lower bound for TypeVars

### ISSUE THREAD

**Issue by poscat0x04** [python/typing#674]
Lower bounds are useful when implementing type safe methods of (classes that are covariant with some type) that take argument of some other type. For example, the method [`updated`](https://www.scala-lang.org/api/current/scala/collection/Seq.html#updated[B>:A](index:Int,elem:B):CC[B]) of scala's `Seq`

reference: #59

**Comment by poscat0x04** [comment]
```
from typing import *

T = TypeVar("T", covariant=True)
V = TypeVar("V")


class List(Generic[T]):
    def updated(self, v: V) -> List[V]:  # V should be a super type of T
        """
        Creates a new list with the first element updated
        """
        pass
```

The argument type cannot be T since it would then occur in the contravariant position.

**Comment by ilevkivskyi** [comment]
This feature appears from time to time in various contexts, see e.g. https://github.com/python/mypy/issues/3737. So I can see the value in it, but I don't think it is a priority ATM.

**Comment by JukkaL** [comment]
Yeah, this would be nice every once in a while. This might help with some `list` methods, such as `copy()`, but only together with F-bounded quantification. I don't understand the `updated` example above. Maybe a more complete example with an implementation would make things clearer.

I agree that this is pretty low priority. We'd probably need to change `typing.TypeVar` to allow specifying a lower bound.

**Comment by erp12** [comment]
I just hit this issue while building some custom data structure types on top of the popular [pyrsistent](https://github.com/tobgu/pyrsistent) library. I was creating custom immutable list and set types that are covariant in the element type. Without lower type bounds, I could not logically implement (at least) the following methods: 
- List: `append`, `prepend`, `concat`
- Set: `add`, `union`, `intersect`
- Both: `reduce`

To provide a better example (not using the invariant `List` type) let's consider the following example using the built-in `Mapping` which is [covariant in the value type](https://github.com/python/cpython/blob/3.9/Lib/typing.py#L1668-L1669). 

```python
from typing import *

K = TypeVar("K")
V = TypeVar("V", covariant=True)
B = TypeVar("B")

class ImmutableMap(Mapping[K, V]):

    def updated(self, key: K, value: B) -> "ImmutableMap[K, ?]":
        """ Returns a new `ImmutableMap` where the `key` is associated with a new `value`."""
        ...
```

The question is: what is the return type of `updated`? There are only a few possible scenarios:

1. `B` is a sub-type of `V` and thus the return type would be `ImmutableMap[K, V]`.
2. `B` is a super-type of `V` and thus the return type would be `ImmutableMap[K, B]`.
3. `B` is some other type with no relationship to `V` and thus the return the must fall back on `ImmutableMap[K, Any]`.

Since we are discussing typed collections here, scenario 3 is very rare and probably irrelevant.

Scenario 1 and 2 are both logical and (in my experience) fairly common. If we could declare `B = TypeVar("B", lower_bound=A)` then it would be possible to support scenarios 1 and 2 in a single method signature without losing any type information.

Without lower bounds, we can only support scenario 1 by annotating `value: V` which is misaligned with the method's logical semantics.

I understand this might stay low-priority. I just wanted to hopefully clear up some of the motivation and document a specific use case that hit this limitation.

**Comment by Galbar** [comment]
I'm having a similar problem. I've written a [dependency injector for python](https://gitlab.com/applipy/applipy_inject/-/blob/master/README.md) (I know the code is horrible) and I'm adding type hints to its code so that it passes mypy checks. I have a method with the signature like this:

```python
...
    def bind(self, types: Tuple[Type[T], ...], instance: S, name: Optional[str] = None, singleton: bool = False) -> None:
```
The idea being that `types` is a tuple of types that `instance` can be represented as (`all(isinstance(instance, t) for t in types)`).

Is there any way this relationship between `T` and `S` can be defined?

**Comment by randolf-scholz** [comment]
I think in principle it should be possible to emulate this by using [slack variables](https://en.wikipedia.org/wiki/Slack_variable), but it seems this usage is not well supported by type checkers at the moment:

```python
from typing import Never, assert_type, reveal_type

class Vec[T]:  # invariant in T
    def _items(self) -> list[T]: ...
    def _append(self, item: T) -> None: ...

def concat_vecs[T1, T2, _T_slack=Never](
    left: Vec[T1],
    right: Vec[T2],
) -> Vec[T1 | T2 | _T_slack]:  # <-- Slack variable makes return a supertype of T1 | T2
    return Vec()

class A: ...
class B: ...
class C: ...

def test_vec_concat(left: Vec[A], right: Vec[B]) -> None:
    # without outer context, the type is precise
    assert_type(concat_vecs(left, right), Vec[A | B])  # ✅️: mypy, pyright, pyrefly, ty, zuban
    # outer context allows widening due to slack variables
    _0: Vec[A | B]     = concat_vecs(left, right) # ✅️: pyright, ty, zuban ❌️: mypy, pyrefly
    _1: Vec[A | B | C] = concat_vecs(left, right) # ✅️: pyright ❌️: mypy, pyrefly, ty, zuban 
    _2: Vec[object]    = concat_vecs(left, right) # ✅️: - ❌️: mypy, pyrefly, pyright, ty, zuban
```

([`pyrefly sandbox`](https://pyrefly.org/sandbox/?project=N4IgZglgNgpgziAXKOBDAdgEwEYHsAeAdAA4CeS4ATrgLYAEALqcROgOZ0Q3G6UN0A5GADcYlADR1UcOGIYB9JsRiTKImKiiLmMADrp9AYyjS4dAGoxDAbQAqAXUR06AYk7phqShAz9WdW31nZ0wYMDp5CAYYGjgAClkoMABKOgBaAD46KAg4BjtHOkJioODQ8PlUYmUsBJgkySiYp1tUzMFcdBgnYsJ9fXK6Q07DVAVRQzg7AEZJWwAmSXlbeTgTQwBrAF4hUUp7ONLnWDAGJ0sbW2n7cSO6bzYACzOLKzt5m-02rIuZugAfALzAERFZrVCbRx3NQMACulHQr0McWS-XQxlMdAAgj0SuiTDI6AAhXF9fGYgDCpLRg2ieXkE3kw3RYziJxevyxN3uECeHLeRPs3w6XUQdzcAHcoo9cLD%2BLLopQhp1ovgGJIGI8YIwdJwzMQ1IZcnpEcFTHJtMo4szRuMrPF2apec9kpJOSDBalXHRAKDkgHg-pw0ZikSRkB7PUOkNRgKAhnWSABesOwGHFdAVYmV6FV-E0UFwErMUtC6FYHEwsO1DFwdHBmzonm8qGwsDgd3kAAZzm8sR77MFnFss7aGfa2WF1Ty%2Bak3P6nGHnZOmInkxg6IAZcgDdCDZEj0dj7em3ZsvcBRJBFP7Q5tY1Hk3Hpyd0%2B9c7oC75G63O7jYbCsY1cZJimiLtvMx7WLg2AAFZWAw-bBNeIy3hMDoTk%2BLovluaSfoGwZ7n%2BP5RouAErsBaIgOIIBytAcAkOQiAgG4ACqDDQFEpB0GAsLoqxnRtgYWBhJxvA0Le6CwjQ2BiHE%2BBOKwDDCnklBiqazgwvCiJgLoIACBJUnKXQwD4AAvtp%2BgUSAv4xqQhDVjQUAUG4AAKUYEbWGA4AQWaQGw8JjBAnRkm4ADKMDas8DDEHAiAAPQxVZsaELwbAxTA6AxZguCTDFzI%2BX5vHpcJSqoJ40DNrA3m8vlAWIrgxAFbR%2BhkJqnRpHscA1XQQ7aQAzIQ0w9WZ6AgMZlEQqxogAGLQDAFBoFgeBEGQI1AA), [ty-playground](https://play.ty.dev/d6ea664d-47a0-4d4b-944f-2b93b435e509), [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByApgG5EgA0UAhgM63kwD68CRVIpR1ANi4kQBQggMY86tKADUiIgNoAVALoAuKFADEmFCWogk1FPlRQFg9eoAmRYFCZIYRCLQAUDHsACUUALQA%2BKB4kWhhFVSgAOijzC2tbJmoENhRLNyIPKgcnNQVvf0IwFCI1KIjhOKgRQpFqZjIRWkUARioFACYqJgUmWnERAGsAXmIyECUXGPUeGxg1GXkFJqUKSah9NAALWelZRTblwTyA%2BeaoAB9TNvO7bt7qAdVVzhgAVxAUHZEXT2ExCSgAIIlaJ-ehQABCwLKoMkAGEoeUbLAiCEmPUmFUUDUYC5psBticAcs1khNgTduClEcCkUVKstAB3BwbMAvfCsxwgSqFRwADxgVBgGyIsAEmEkCE4ImCQgsNHojH4bBcmOxaNkrjxApJZM8VEJ10p3k0UEAoOSAeD%2B1BBEHAqAg4OstnaHTYeLbRVQAF4vABGhnpUA55G5RiI-JoPB4YAZkiZ1hQqAwlheIpgYCgdwGUF0%2BmoPumtFWTAADHNdgDDUo5VBBiG1fVNTMOKStt4tBa1PbHdr4F7fYYoIAZcktUGt9udnGAbqLTTL8grF3B11hVdrqtq6oauKbOtbJo7UC7Lfww6tNonrvdvag3r9HyLbTncjAPoAVrIYFWLGvqhuG9v8WbXV9xHHwhxHMd3S7S8J2PQV3VvQwgA))
